servicemanager/servicemanager.go

```diff
@@ -36,34 +46,211 @@ const maxConcurrentActions = 10
  * Public
  *********************************************************************************************************************/
 
 // New creates new launcher object.
-func New(config *config.Config) (sm *ServiceManager, err error) {
+func New(config *config.Config, serviceInfoProvider ServiceInfoProvider) (sm *ServiceManager, err error) {
 	sm = &ServiceManager{
-		servicesDir:   config.ServicesDir,
-		actionHandler: action.New(maxConcurrentActions),
+		servicesDir:         config.ServicesDir,
+		downloadDir:         config.DownloadDir,
+		actionHandler:       action.New(maxConcurrentActions),
+		idsPool:             uidgidpool.NewGroupIDPool(),
+		serviceInfoProvider: serviceInfoProvider,
 	}
 
 	if err = os.MkdirAll(sm.servicesDir, 0o755); err != nil {
 		return nil, aoserrors.Wrap(err)
 	}
 
+	services, err := sm.serviceInfoProvider.GetAllServices()
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	for _, service := range services {
+		if err = sm.idsPool.AddID(service.GID); err != nil {
+			log.Errorf("Can't add service UID/GID to pool: %s", err)
+		}
+	}
+
 	return sm, nil
 }
 
+// InstallService install service to the system.
+func (sm *ServiceManager) InstallService(
+	newService ServiceInfo, imageURL string, fileInfo image.FileInfo) (err error) {
+	return <-sm.actionHandler.Execute(newService.ServiceID,
+		func(id string) error {
+			return sm.doInstallService(newService, imageURL, fileInfo)
+		},
+	)
+}
+
 /**********************************************************************************************************************
  * Private
  *********************************************************************************************************************/
+
+func (sm *ServiceManager) doInstallService(newService ServiceInfo, imageURL string, fileInfo image.FileInfo) error {
+	serviceFromDB, err := sm.serviceInfoProvider.GetService(newService.ServiceID)
+	if err != nil && !errors.Is(err, ErrNotExist) {
+		return aoserrors.Wrap(err)
+	}
+
+	var isServiceExist bool
+
+	if err == nil {
+		isServiceExist = true
+
+		if newService.AosVersion <= serviceFromDB.AosVersion {
+			return ErrVersionMismatch
+		}
+	}
+
+	// download and unpack
+	urlVal, err := url.Parse(imageURL)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	var serviceImage string
+
+	if urlVal.Scheme != "file" {
+		if serviceImage, err = image.Download(context.Background(), sm.downloadDir, imageURL); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		defer os.RemoveAll(serviceImage)
+	} else {
+		serviceImage = urlVal.Path
+	}
+
+	if err = image.CheckFileInfo(context.Background(), serviceImage, fileInfo); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// create install dir
+	newService.ImagePath, err = ioutil.TempDir(sm.servicesDir, "")
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = imageutils.UnpackTarImage(serviceImage, newService.ImagePath); err != nil {
```

Comment:
Let's move `imageutils.UnpackTarImage` to common.

Reply:
will be done in next mrs